kytos/core/link.py

```diff
@@ -130,56 +131,50 @@ def available_tags(self):
         return (self.endpoint_a.is_tag_available(tag) and
                 self.endpoint_b.is_tag_available(tag))
 
-    def get_next_available_tag(self):
+    def get_next_available_tag(self, controller, tag_type: str = '1') -> TAG:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
             # Copy the available tags because in case of error
             # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
+            available_tags_a = deepcopy(self.endpoint_a.get_available_tags())
+            available_tags_b = deepcopy(self.endpoint_b.get_available_tags())
+            intersection_tags = Interface.range_intersection(available_tags_a,
+                                                             available_tags_b)
+            for tag_range in intersection_tags:
+                for tag in range(tag_range[0], tag_range[1]+1):
+                    # Tag already in use. Try another tag.
+                    if not self.endpoint_a.use_tags([tag, tag]):
+                        continue
+
+                    # Tag already in use in B. Mark the tag as available again.
+                    if not self.endpoint_b.use_tags([tag, tag]):
+                        self.endpoint_a.make_tags_available([tag, tag])
+                        continue
+
+                    # Tag used successfully by both endpoints. Returning.
+                    self.endpoint_a.notify_link_available_tags(controller)
+                    self.endpoint_b.notify_link_available_tags(controller)
+                    return TAG(int(tag_type), tag)
 
             raise KytosNoTagAvailableError(self)
 
-    def make_tag_available(self, tag):
+    def make_tag_available(
+        self,
+        controller,
+        tag: int,
+        tag_type: str = '1'
+    ) -> (bool, bool):
         """Add a specific tag in available_tags."""
-        if not self.is_tag_available(tag):
-            self.endpoint_a.make_tag_available(tag)
-            self.endpoint_b.make_tag_available(tag)
-            return True
-        return False
+        result_a = self.endpoint_a.make_tags_available([tag, tag], tag_type)
```

Comment:
This should also use a Lock, `_get_available_vlans_lock` is a singleton lock, it might be too much, but still it'll need a lock.

Reply:
check out this comment https://github.com/kytos-ng/kytos/pull/409/files#r1309215356, it's partly related too

Reply:
tag_type changed to string.